### PR TITLE
[algs][0] Protobuf types and key generation

### PR DESCRIPTION
**Pull request by nklaassen**
This PR begins the implementation of [RFD 136 - Modern Signature Algorithms](https://github.com/gravitational/teleport/blob/master/rfd/0136-modern-signature-algorithms.md). It defines the protobuf types for configuring the current suite via the `cluster_auth_preference`, and introduces partial definitions for the suites and basic functions for generating key material.

Further PRs will build on this to support generating keys matching the currently configured algorithm suite for each key purpose.

### CODE REVIEW

**Review by probakowski — APPROVED** [review]

**Review by rosstimothy — COMMENTED** [review]
If we aren't going to ship this until v17 would it be less work if we dropped the backports and omitted the `-dev` suites and subsequent deprecation process?

**Comment by nklaassen** [comment]
> If we aren't going to ship this until v17 would it be less work if we dropped the backports and omitted the `-dev` suites and subsequent deprecation process?

A bit less work later I guess, and if we end up having to use the enum values somewhere we won't be skipping 2-4. I switch to just using the v1 suite names

File: lib/cryptosuites/suites.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+	ECDSAP256
+	// Ed25519 represents Ed25519 keys.
+	Ed25519
+
+	// algorithmMax is 1 greater than the last valid algorithm, used to test that an algorithm is valid.
+	algorithmMax
+)
+
+// Suite defines the cryptographic signature algorithm used for each unique key purpose.
+type Suite map[KeyPurpose]Algorithm
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply lib/cryptosuites/suites.go:104]: I can actually unexport this

File: lib/services/authentication_test.go [hunk 2]
```diff
@@ -82,3 +82,52 @@ func TestValidateLocalAuthSecrets_passwordHash(t *testing.T) {
 	assert.True(t, trace.IsBadParameter(err),
 		"ValidateLocalAuthSecrets returned err=%v (%T), want BadParameter", err, trace.Unwrap(err))
 }
+
+func TestSignatureAlgorithmSuiteRoundtrip(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply lib/services/authentication_test.go:87]: added tests for all other cases, and simplified to only cover float64 which json.Unmarshal will always use for numbers

File: api/types/signaturealgorithmsuite.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		} else {
+			return trace.BadParameter("SignatureAlgorithmSuite invalid value: %s", str)
+		}
+	}
+	return nil
+}
+
+// UnmarshalJSON marshals a SignatureAlgorithmSuite value to JSON.
+func (s *SignatureAlgorithmSuite) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + s.toString() + `"`), nil
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply api/types/signaturealgorithmsuite.go]: switched to `json.Marshal(s.toString())` which is quite a bit less questionable I guess

[the remark this answers is not in the document]
Comment (nklaassen) [reply api/types/signaturealgorithmsuite.go]: Oh TIL thanks

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**